### PR TITLE
Filepath select

### PR DESCRIPTION
**Pull request by anwarhahjjeffersongeorge**
Hi. I tried to make a file path selector that:
- Is its own prompt (`PathSelect`) with optional configuration (`PathSelectPrompt`) gated behind the `path` feature
- Is OS-agnostic,
- Uses the [`fs_err`](https://lib.rs/crates/fs-err) crate where possible for more detailed diagnostics, 
- Allows starting from a custom file path `start_path_opt` (defaulting to current process directory or root path if current process directory is unavailable)
- Allows selection of one or multiple files with configuration option `select_multiple`,
- Allows selecting (with `select_multiple` enabled) files from more than one directory, 
- Allows filtering of selectable file types by directory, extension, or a combination of multiple filters `selection_mode`, 
- Allows following/showing symbolic links with configuration option `show_symlinks`.
- Allows showing hidden files with configuration option `show_hidden`,
- Uses Rust's `std::path` and `std::ffi::OsStr` for portability instead of trying to figure out whether to convert paths to strings, and
- Includes an example (path_select.rs).

### Deficiencies
It does not:
- Deal with permissions (it's possible, but I didn't know if it was a good idea),
- Use existing prompts, although it is based heavily on the `MultiSelect` prompt code, 
- Handle formatting very aesthetically,
- Support certain types of hidden files as I neither have a windows machine for testing nor understand how it works :(, 
- Support user-provided custom option filtering beyond selecting acceptable file types using the `PathSelectionMode` type , 
- Support user-provided validation functions,
-  Show relative paths  (there are crates to do this, but they seem to have some [open issues](https://github.com/Manishearth/pathdiff/issues)), or
- Have very many tests. I didn't know how you wanted tests written. For what it's worth, everything passes when I run `cargo test --all-features-`. It seems to work when I tried it on Ubuntu 23.04 and Mac OS 10.15.7.

I am opening this PR mainly for feedback. I know the contributor's guide said to do the discussion stage first, but I needed this feature for a project and did not see that detail until after I made it. I saw some existing issues (#124, #63), and tried to go from there, favoring the strategy of 63 ( a new prompt type) with the features of 124 (as detailed above). 

With all that in mind, I understand if this needs more work.

### CODE REVIEW

**Comment by mikaelmello** [comment]
I really like this. And thank you for the attention to detail on the PR description lol.

Hopefully I'll find some time tomorrow for a full review.

One important point that I'd bring up is that recently (like yesterday maybe) I've refactored prompts. Far from a revamp but significant enough that you might want to get ahead in porting this PR already. Honestly, I can also do that myself if you'd prefer, whatever works.

**Comment by anwarhahjjeffersongeorge** [comment]
I gave refactoring the prompts a shot. For now, I'll chill out on this until whenever you get a chance to look and let me know what needs improvement.

File: inquire/src/prompts/path_select/action.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    MoveToStart,
+    /// Moves the cursor to the end of the list.
+    MoveToEnd,
+    /// Toggles the selection of the current option.
+    ToggleCurrentOption,
+    /// Selects all options.
+    SelectAll,
+    /// Deselects all options.
+    ClearSelections,
+    /// Navigates deeper in file tree
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/action.rs:31]:
```suggestion
    /// Navigates deeper in file tree when current selection is a directory
```

wdyt?

File: inquire/src/prompts/path_select/action.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    NavigateDeeper,
+    //// Navigated higher in file tree
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/action.rs:33]:
```suggestion
    /// Navigates higher in file tree.
    ///
    /// No-op if on a root directory.
```

wdyt? I know it's pretty obvious but I like the "being very explicit" approach

File: inquire/src/prompts/path_select/action.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    NavigateHigher,
+}
+
+impl InnerAction<PathSelectConfig> for PathSelectPromptAction {
+    fn from_key(key: Key, config: &PathSelectConfig) -> Option<Self> {
+        if config.vim_mode {
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/action.rs:39]: I have to do some proper vim mode, because the current way is kind of bad. e.g. on vim_mode there is no way for users to type `k` or `j` in the input.

Let's not use vim_mode on this new prompt for now, and once I fix it for the others I can add it here.

File: inquire/src/prompts/path_select/mod.rs [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+#[derive(Clone, Eq, PartialEq)]
+pub enum PathSelectionMode<'a> {
+    /// The user may pick a file with the given (optional) extension
+    File(Option<&'a str>),
+    /// The user may pick a directory
+    Directory,
+    /// The user may pick multiple paths
+    Multiple(Vec<PathSelectionMode<'a>>),
+}
+impl<'a> Default for PathSelectionMode<'a> {
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/mod.rs:37]:
```suggestion
}

impl<'a> Default for PathSelectionMode<'a> {
```

nit

File: inquire/src/prompts/path_select/mod.rs [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+    fmt, fs,
+    ops::Deref,
+    path::{Path, PathBuf},
+};
+
+/// Different path selection modes specify what the user can choose
+#[derive(Clone, Eq, PartialEq)]
+pub enum PathSelectionMode<'a> {
+    /// The user may pick a file with the given (optional) extension
+    File(Option<&'a str>),
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/mod.rs:31]: @LeoRiether, is using `Cow<'a, str>` here a good idea? From your comments on that another issue

File: inquire/src/prompts/path_select/mod.rs (hunk 4 above)
Comment (mikaelmello) [reply inquire/src/prompts/path_select/mod.rs:37]: (also applies to all the other pairs without a new line in the middle.

File: inquire/src/prompts/path_select/mod.rs [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+}
+impl<'a> Default for PathSelectionMode<'a> {
+    fn default() -> Self {
+        Self::Directory
+    }
+}
+
+/// Path with cached information
+#[derive(Clone, Debug, Hash)]
+pub struct PathEntry {
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/mod.rs:45]: I think the amount of PathEntry impls warrant a new module just for PathEntry, living alongside this current file.

File: inquire/src/prompts/path_select/mod.rs [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+    }
+
+    /// Is this path entry for a symlink?
+    pub fn is_symlink(&self) -> bool {
+        self.symlink_path_opt.is_some()
+    }
+}
+/// Prompt for choosing one or multiple files.
+///
+/// The user can
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/mod.rs:144]: :)

File: inquire/src/prompts/path_select/mod.rs [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+
+    /// Function that formats the user input and presents it to the user as the final rendering of the prompt.
+    pub formatter: MultiOptionFormatter<'a, PathEntry>,
+
+    /// Whether the current filter typed by the user is kept or cleaned after a selection is made.
+    pub keep_filter: bool,
+
+    /// RenderConfig to apply to the rendered interface.
+    ///
+    /// Note: The default render config considers if the NO_COpubLOR environment variable
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/mod.rs:186]:
```suggestion
    /// Note: The default render config considers if the NO_COLOR environment variable
```

File: inquire/src/prompts/path_select/mod.rs [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+    /// use inquire::PathSelect;
+    /// use std::path::Path;
+    ///
+    /// assert!(PathSelect::is_path_hidden_file(Path::new("/ra/set/.nut")));
+    /// assert!(!PathSelect::is_path_hidden_file(Path::new("/ra/set/nut")));
+    /// assert!(PathSelect::is_path_hidden_file(Path::new(".maat")));
+    /// assert!(!PathSelect::is_path_hidden_file(Path::new("maat")));
+    ///
+    /// ```
+    pub fn is_path_hidden_file(t: T) -> bool {
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/mod.rs:338]: I think this should be inside PathSelectPrompt and not on PathSelect (the prompt builder API), if I got it right this is only used within the impl and not supposed to be used by the user right?

File: inquire/src/prompts/path_select/prompt.rs [hunk 10]
```diff
@@ -0,0 +1,567 @@
...
+                let default_path = default_item.as_ref();
+                if !default_path.exists() {
+                    Err(InquireError::InvalidConfiguration(format!(
+                        "Specified default path `{default_path:?}` does not exist"
+                    )))
+                } else {
+                    Ok(())
+                }
+            })?;
+        }
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/prompt.rs:64]: I think the behavior for default will be better if it's "selected if exists", without raising an error if not.

The why is that, thinking from the perspective of the developer using inquire, they're gonna have to make a very strong assumption on this path existing on the _user's machine_, and if they don't exist by any reason, this will be an unrecoverable error for the user.

This will force everyone using that feature to check if the path exists before creating this default list, which will be extra boilerplate we can remove if the goal is "if this path exists, select it".

In other words, if a path existing is so important for me that the prompt becomes irrelevant (should error out) if it doesn't exist, I as a developer likely want to make that assertion myself, with my own error handling, rather than delegating this check to inquire.

I'm just going through the code but I still want to write a few examples and see how the prompt feels UX-wise before making any assertions, so once I actually use this prompt (as you probably are already), I might change my idea here if I'm speaking non-sense.

**Review by mikaelmello — COMMENTED** [review]
I took the first half hour I had to not keep you waiting too much!

I still haven't reviewed much of the prompt implementation itself, nor could tinker with PathSelect as an user, but so far it looks pretty solid!!! Thanks for your fantastic work

File: inquire/src/prompts/path_select/mod.rs (hunk 5 above)
Comment (LeoRiether) [reply inquire/src/prompts/path_select/mod.rs:31]: There are pros and cons. `Cow` in some cases does provide easier lifetime handling, but it also adds some complexity to the API, like the need to write `File(Some("jpg".into()))` instead of `File(Some("jpg"))` and some users saying "what the hell is a Cow?"

Maybe it's worth it? I didn't find clear guidelines about this online, but I'll look into it again later to try and come up with a more definitive answer :)

Comment (LeoRiether) [reply inquire/src/prompts/path_select/mod.rs:31]: Also, I feel like it would be nice to have a more flexible `PathSelectionMode`. I don't see an easy way, for example, to integrate the [ignore](https://crates.io/crates/ignore) crate to list files respecting `.gitignore`. If I have any ideas of how to actually implement this more flexible API, I'll make sure to post them.

File: inquire/src/prompts/path_select/mod.rs (hunk 9 above)
Comment (mikaelmello) [reply inquire/src/prompts/path_select/mod.rs:235]: Just the function

File: inquire/src/prompts/path_select/action.rs (hunk 3 above)
Comment (mikaelmello) [reply inquire/src/prompts/path_select/action.rs]: If the config won't change the behavior, it might be better to remove it entirely so it doesn't confuse the users

File: inquire/src/prompts/path_select/mod.rs (hunk 5 above)
Comment (anwarhahjjeffersongeorge) [reply inquire/src/prompts/path_select/mod.rs:27]: > . I don't see an easy way, for example, to integrate the [ignore](https://crates.io/crates/ignore) crate to list files respecting `.gitignore`.

Hiya. I'm not sure what you're asking for here. Would it help if I tried to add globs or some kind of name-matching/name-excluding patterns? I looked at the `ignore` crate, but I don't know if adding that dependency would be better than just using the `regex` crate. Are you wanting to respect `.gitignore` files in the ancestry of paths too?

File: inquire/src/prompts/path_select/prompt.rs (hunk 10 above)
Comment (anwarhahjjeffersongeorge) [reply inquire/src/prompts/path_select/prompt.rs]: That makes sense. I still ended up filtering out non-existent supplied paths because it seemed to be the easiest way to avoid errors that would come up in rendering.

File: inquire/src/prompts/path_select/mod.rs [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+    /// let b = PathBuf::from("/maat/nut.rs");
+    /// ans.push(ListOption::new(3, &b));
+    /// assert_eq!(String::from("/ra/set/nefer.rs, /maat/nut.rs"), formatter(ans.as_slice()));
+    ///
+    /// let c = PathBuf::from("ptah.rs");
+    /// ans.push(ListOption::new(7, &c));
+    /// assert_eq!(String::from("/ra/set/nefer.rs, /maat/nut.rs, ptah.rs"), formatter(ans.as_slice()));
+    /// ```
+
+    pub const DEFAULT_PATH_FORMATTER: MultiOptionFormatter<'a, T> = &|ans| {
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/mod.rs:122]:
```suggestion
    /// ```
    pub const DEFAULT_PATH_FORMATTER: MultiOptionFormatter<'a, T> = &|ans| {
```

File: inquire/src/prompts/path_select/mod.rs [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+    pub const DEFAULT_SELECT_MULTIPLE: bool = false;
+
+    /// Default visual divider value.
+    pub const DEFAULT_DIVIDER: &'a str = "-----";
+
+    /// Creates a [PathSelect] with the provided message and options, along with default configuration values.
+    pub fn new(message: &'a str, start_path_opt: Option<T>) -> Self {
+        Self {
+            message,
+            start_path_opt,
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/mod.rs:190]: If it's optional, let's make users set it via the builder and allow the happy path to be frictionless

```suggestion
    pub fn new(message: &'a str) -> Self {
        Self {
            message,
            start_path_opt: None,
```

File: inquire/src/prompts/path_select/mod.rs [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+    /// Parses the provided behavioral and rendering options and prompts
+    /// the CLI user for input according to the defined rules.
+    ///
+    /// Returns a [`ListOption`](crate::list_option::ListOption) containing
+    /// the index of the selection and the owned object selected by the user.
+    pub fn raw_prompt(self) -> InquireResult<Vec<ListOption<PathEntry>>> {
+        let terminal = get_default_terminal()?;
+        let mut backend = Backend::new(terminal, self.render_config)?;
+        self.prompt_with_backend(&mut backend)
+    }
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/mod.rs:357]: My reasoning for raw_prompts back then was that some devs might want to have the index so they could reference it to the vectors they created the prompt with. Since the options here will be dynamically retrieved, I don't see much value (or at all) in returning indexes here, so let's just remove these methods.

```suggestion
```

File: inquire/src/prompts/path_select/path_entry.rs [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+        let size_formatting = if cfg!(windows) {
+            humansize::WINDOWS
+        } else {
+            humansize::DECIMAL
+        };
+        let size_string = self.size.format_size(size_formatting); 
+        if let Some(symlink_path) = self.symlink_path_opt.as_ref() {
+            write!(f, "{} -> {path} ({size_string})", symlink_path.to_string_lossy())
+        } else if self.is_dir() {
+            write!(f, "(dir) {path} ({size_string})")
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/path_entry.rs:72]: Wouldn't this always be 0?

File: inquire/src/prompts/path_select/modes.rs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+#[derive(Clone, Eq, PartialEq)]
+pub enum PathSelectionMode<'a> {
+    /// The user may pick a directory.
+    Directory,
+    /// The user may pick a file with the given (optional) extension.
+    File(Option<&'a str>),
+    /// The user can set gitignore rules from a file 
+    /// The user may pick multiple paths
+    Multiple(Vec<PathSelectionMode<'a>>),
+}
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/modes.rs:17]: I feel like this can be made simpler for end-users. On a first glance, it's not quite intuitive how this selection mode behaves, with the recursive Multiple warranting more attention.

wdyt of:

accept enum, like JS's accept attribute: All, Only(String), Any(Vec<String>)

Then PathSelectionMode
- Directory
- File(Accept)
- FileOrDirectory(Accept)

I still don't fee this is the best solution, but it feels more intuitive

File: inquire/src/prompts/path_select/modes.rs [hunk 16]
```diff
@@ -0,0 +1,50 @@
+//! Prompt modes 
+//! 
+//! Everything here should be exposed to the (library) user
+use strum::{Display, EnumCount, FromRepr};
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/modes.rs:4]: I think that, for the single purpose of enriching PathSortingMode, it is better to implement these methods ourselves than pulling in the dependency

File: inquire/src/prompts/path_select/modes.rs [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    /// The user may pick a file with the given (optional) extension.
+    File(Option<&'a str>),
+    /// The user can set gitignore rules from a file 
+    /// The user may pick multiple paths
+    Multiple(Vec<PathSelectionMode<'a>>),
+}
+
+impl<'a> Default for PathSelectionMode<'a> {
+    fn default() -> Self {
+        Self::Directory
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/modes.rs:21]: I think defaulting to File makes more sense here, specially when we think that picking directories was barely supported on browsers years ago

File: inquire/src/prompts/path_select/prompt.rs [hunk 18]
```diff
@@ -0,0 +1,576 @@
...
+            current_path,
+            sorting_mode,
+            ..
+        } = self;
+
+        let prompt = &format!(
+            r#"{message} 
+- Current directory: {} 
+- Sorted by {sorting_mode} (Press tab to change). 
+- Type to filter: "#,
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/prompt.rs:430]:
```suggestion
- Type to filter:"#,
```

File: inquire/src/prompts/path_select/prompt.rs [hunk 19]
```diff
@@ -0,0 +1,576 @@
...
+                    }
+                }
+            }
+
+            PathSelectPromptAction::NavigateHigher => {
+                if let Some(parent) = current_path.parent() {
+                    *current_path = parent.to_path_buf();
+                    *data_needs_refresh = true;
+                    *cursor_index = 0;
+                }
```
Comment (mikaelmello) [inline inquire/src/prompts/path_select/prompt.rs:526]: I got this error when going up and up, not sure why as it seems you put the handling into place, but worth checking out
```
IO(
    Os {
        code: 2,
        kind: NotFound,
        message: "No such file or directory",
    },
)
```

**Comment by mikaelmello** [comment]
I do feel bad about dumbing a lot of comments, but it is a very large feature :). I want it to be as good as possible before releasing, which today is basically the same as merging.

I was thinking, we should break this down in smaller chunks and get the thing going more gradually, we can merge but not expose this on the published crate until we gradually get this to the end state.

I think this would also allow us to work in parallel.

Also, obligatory sorry for taking too long on this round of reviews

**Comment by Barre** [comment]
Hi,

I am also interested in this.

Did this go anywhere? Any way I can help?

**Comment by anwarhahjjeffersongeorge** [comment]
@mikaelmello  @Barre  Hey there, I missed this before, but I'll take a look at these next weekend.

File: inquire/src/prompts/path_select/path_entry.rs (hunk 14 above)
Comment (anwarhahjjeffersongeorge) [reply inquire/src/prompts/path_select/path_entry.rs:72]: As I coded it, yes. I wrote it using the `size_string` here for consistency to avoid having to update the 0 value in two places. 

Using of the 0 value for directories was a workaround because for me on Linux, directory sizes always show up as 4096. I am not sure if there are OSes where the `Metadata::len` function returns something more correct, but I wanted to leave it open to the possibility in case you decide to provide an actual accurate size with an OS-specific compiler flag.

File: inquire/src/prompts/path_select/prompt.rs (hunk 19 above)
Comment (anwarhahjjeffersongeorge) [reply inquire/src/prompts/path_select/prompt.rs:526]: I am not able to replicate this on Ubuntu Linux, but it was indeed a persistent problem before. Can you please let me know what OS you were on so I can try to write a test or something and figure it out?

File: inquire/src/prompts/path_select/modes.rs (hunk 17 above)
Comment (anwarhahjjeffersongeorge) [reply inquire/src/prompts/path_select/modes.rs]: No problem

File: inquire/src/prompts/path_select/modes.rs (hunk 15 above)
Comment (anwarhahjjeffersongeorge) [reply inquire/src/prompts/path_select/modes.rs]: Sure, I suppose it was redundant the way it was. I tried implementing an inner  enum (`PathFilter`) that can accept or deny based on file extensions, stems, and matching functions. I also added some documentation tests for the different modes and tried to tighten up the documentation.

File: inquire/src/prompts/path_select/mod.rs [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+    terminal::get_default_terminal,
+    type_aliases::Filter,
+    ui::{Backend, MultiSelectBackend, RenderConfig},
+    InquireError,
+};
+use std::path::Path;
+
+/// Prompt for choosing one or multiple files.  
+///
+/// The user can
```
[the remark this answers is not in the document]
Comment (anwarhahjjeffersongeorge) [reply inquire/src/prompts/path_select/mod.rs]: Let me know what you think. One thing I did not document was the divider. It is currently unused because I couldn't figure out how to make it work with the prompt — basically, I thought it would be less confusing with a visual divider between the current directory's entries and items selected from previous directories if any. Only I don't know how to insert a divider into the list or if that's possible, so maybe you can suggest something?

**Comment by anwarhahjjeffersongeorge** [comment]
I was thinking, we should break this down in smaller chunks and get the thing going more gradually, we can merge but not expose this on the published crate until we gradually get this to the end state.
> 
> I think this would also allow us to work in parallel.

Just let me know how you'd like me to reorganize it that would work for you.

> Also, obligatory sorry for taking too long on this round of reviews

No problem. It's been like 110+ degrees where I live, so haven't exactly been rushing to turn the computer on and heat up the house.

**Comment by DaRacci** [comment]
any progress with this?

**Comment by anwarhahjjeffersongeorge** [comment]
> any progress with this?

@mikaelmello Said above they wanted me to reorganize or break this down but I'm not sure what that meant, so I'm waiting on that before doing more with this.

**Comment by mikaelmello** [comment]
hey, I know I've been really dropping the ball on being responsive throughout inquire, life getting in the way as usual

I really really want to see this merged, the main point holding it back is that this is a very significant prompt to offer, counter-intuitively.

it's one of those things I'd like to polish and re-polish it before releasing, making sure the API is very intuitive and everything

**Comment by gswebspace** [comment]
Ran into this requirement/feature. Any progress ?
